internal/ingress/controller/store/store.go

```diff
@@ -358,6 +358,178 @@ func New(namespace string, configmap, tcp, udp, defaultSSLCertificate string,
 	store.informers.Secret.AddEventHandler(secrEventHandler)
 	store.informers.Service.AddEventHandler(serviceEventHandler)
 
+	pluginEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			plugin := obj.(*pluginv1.KongPlugin)
+			if !class.IsValid(&plugin.ObjectMeta) {
+				a, _ := parser.GetStringAnnotation(class.IngressKey, &plugin.ObjectMeta)
+				glog.Infof("ignoring add for plugin %v based on annotation %v with value %v", plugin.Name, class.IngressKey, a)
+				return
+			}
+
+			updateCh.In() <- Event{
+				Type: ConfigurationEvent,
+				Obj:  obj,
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			plugin, ok := obj.(*pluginv1.KongPlugin)
+			if !ok {
+				// If we reached here it means the ingress was deleted but its final state is unrecorded.
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					glog.Errorf("couldn't get object from tombstone %#v", obj)
+					return
+				}
+				plugin, ok = tombstone.Obj.(*pluginv1.KongPlugin)
+				if !ok {
+					glog.Errorf("Tombstone contained object that is not a KongPlugin: %#v", obj)
+					return
+				}
+			}
+			if !class.IsValid(&plugin.ObjectMeta) {
+				glog.Infof("ignoring delete for plugin %v based on annotation %v", plugin.Name, class.IngressKey)
+				return
+			}
```

Comment:
There seems to be a lot of duplication of code here.
Can we create a method for checking the validity of the resource and then reuse that across all 3 cache handlers?

Reply:
I've created 3 new methods in `class` package:
- `CanAddResource`
- `CanDeleteResource`
- `CanUpdateResource`

I've also renamed the event handler `annotatedCrdEventHandler` which is now common to all CRD that supports annotations.

Reply:
The new code seems very odd in the way it is organized.

Organizing the code as follows:

```go
		AddFunc: func(obj interface{}) {
                   var p *pluginv1.KongPlugin
                   if p, ok := obj.(*pluginv1.KongPlugin); !ok {
                     // err  
                   }
		   objectMeta = &p.ObjectMeta			
                   if ok := class.IsValid(objectMeta); !ok
                     return
		    }
		    updateCh.In() <- Event{
		       Type: ConfigurationEvent,
		       Obj:  obj,
		     }
		},
```

This helps in a few ways:
- Avoid duplicate logic per type
- IsValid is the only method that is called, making it easier to test and introduces changes in future.

Reply:
@hbagdi do you prefer 1 event handler per resource because the code you've suggested means this... I tried to keep only one handler anyway I can fall back to 3 handlers and have your last suggestion.

Reply:
I think 3 handlers are okay, since we need to type cast. But we should keep the hanlders small and have minimal logic as possible.

Reply:
Ok pushed.